### ISSUE TITLE
Build Well gbXML Updates

### ISSUE THREAD

**Issue by bwelle** [ladybug-tools/spider#27]
Discussion forum for gbXML export functionality.

**Comment by bwelle** [comment]
I see you've put a long list of gbXML objects in the gbXML settings tab. Our gbXML will only require the minimum number of objects to define geometry only. All other information is handled outside the gbXML. The objects we will use includes..

openingTypeEnum
BuildingStorey
Azimuth
Area
ObjectId
surfaceTypeEnum

There are probably some more, we will look to Chien Si to identify them. We will take a box shape and generate what the gbXML should look like. We can develop gbXML export to match this file.

**Comment by theo-armour** [comment]
@bwelle 

> I see you've put a long list of gbXML objects in the gbXML settings tab. 

It is not just a list of objects, it is THE list of gbXML objects - all scraped at every run time from the canonical schema att https://github.com/GreenBuildingXML/.

For sure, it is overkill for here and now, but it points the way toward online, real-time gbXML validation.

I look forward to a good list of objects to be included.

BTW, Chien Si has not yet accepted the invite to collaborate - so may not be following the issues posted here. See:

https://github.com/ladybug-tools/spider/settings/collaboration

**Comment by mdengusiak** [comment]
I will be happy to test your  gbXML on my side as well, please let me know when you manage to get first files.

**Comment by bwelle** [comment]
The validator is pretty crude.  But for the test cases that it does have, it works pretty well.

1.	https://github.com/GreenBuildingXML/ValidatorPhase1/wiki.  It explains what the validator checks :
1.	against an XSD
2.	is "similar" to the gbXML on the server that represents this test case
3.	phase 3 has not started, but the idea is to get a handle on people's intent.
2.	http://gbxml.org/validator/Pages/TestPage.aspx (you have to redraw a test case, in your tool, export it, and push it into the validator.
3.	https://github.com/GreenBuildingXML/gbXMLReadWriteSDK_CSharp/wiki (a good place to understand what gbxml is all about.  this code repo I wrote which is in C# unfortunately.  But this wiki has a nice summary of all required fields.  It hasn't been updated, but it should be 98% correct.).
1.	For the Space element, I would relax the constraint for now that you need all this stuff.  If you just have a space, ID, area, and volume, then that should be all you need to get started for now.  I'm pretty sure that OpenStudio doesn't care about the rest.

**Comment by theo-armour** [comment]
I am not having your luck. ;-(

I have downloaded three models from here:

https://github.com/GreenBuildingXML/Sample-gbXML-Files

I have run different test with different schemas. 

All are failing.

**Comment by theo-armour** [comment]
@chiensiTB 

Can you help me get at least one test/schema/file combo working?

**Comment by chiensiTB** [comment]
Hey guys-

I'm not sure you are using the validator the right way.  I'll have to
explain it to you more clearly.

**Comment by theo-armour** [comment]
I downloaded several of the sample files to my hard drives.

results for gbXMLStandard Single Family Residential 2016.xml

[image: Inline image 1]

**Comment by theo-armour** [comment]
same file / schema 6

[image: Inline image 1]

**Comment by chiensiTB** [comment]
Theo-

I'd stop.  I'm pretty sure you are misinterpreting what the validator
does.  What the validator does is compare your gbxml to a gbxml
representation of the SAME BUILDING.  Think of it as a benchmark testbed as
opposed to a pure validation tool.

So if you select Test 3 in the dropdown menu, you must upload YOUR VERSION
of Test 3 (in other words, you have to draw it) and test YOUR version vs.
the version that is already on the server.

You can't simply upload any gbxml file you want.

Does this make sense?

**Comment by theo-armour** [comment]
For example these two files?

https://github.com/GreenBuildingXML/openstudio-gbxml-validation/tree/master/gbxml_test_suite/tests/output/test_case_3

**Comment by bwelle** [comment]
Summary Notes gbXML meeting and merging of Build Well Space/Zoning Integration and Build Well gbXML Updates issues:

Theo to add vertice box for all vertices in Build Well. We will not be handling complex surfaces or geometry in this project so doing this should clutter things.

For a valid gbXML file, we need the required elements identified here:

[https://github.com/GreenBuildingXML/gbXMLReadWriteSDK_CSharp/wiki]

We primarily need to support spaces and surfaces and openings. For our purposes, surface geometry is the same as space boundary geometry. Chien Si, can we exclude space boundaries and just use surfaces?

Surface ID attributes will use the following logic: Elements are either Slab, Wall, Window, or Shading. Just use sequential integers for the elements tagged at the end of these main headings, for example. Slab.1, Slab.2, Wall.1, Wall.2, Wall.3, etc. It doesn’t matter which surface you start with 1 or the next one you go to for 2, etc.

Space ID attributes will be Space.1, Space.2, Space.3, etc. There is a specific space (zone, same term for our purposes) that should start with 1, then an order how to identify Space.2, Space.3, etc. See [SpaceTypeZoneAssignment.pdf](https://perkinswill.box.com/s/f3208f9m1wsg0vjx1vbmsjr4plywpf1n). Convention for zone naming is this: Zone 1 is always lower center zone, which usually corresponds to the zone adjacent to Length 1 (or X1 in the attached [Footprint Shapes pdf](https://perkinswill.box.com/s/oeh672p80fji3p11krjluxa8xpltn91k))). In the picture attached, Zone 1 is at the top since the shape has been rotated 180 degrees. Once Zone 1 is known, the zones are numbered clockwise until all outer zones have been named. The core is the last zone (for the donut shape, we would have an extra loop of “outer zones”, but forget that for now). When there is more than 1 floor (e.g. 2), the zone numbering continues sequentially, starting with the zone directly above the original Space.1. In other words, an L-shape has 7 spaces. If 2 floors, Space.8 will be directly above Space.1, etc.
We will ignore space types assigned to zones in the GUI. This we can handle in the DSC engine.
Object Types need to be assigned to each surface. The object types we need to support are:

a.	Exterior Wall
b.	Interior Wall
c.	Window
d.	Overhang
e.	Fin
f.	RoofSlab
g.	BaseSlab
h.	FloorSlab
i.	AdjacentBuilding

Convention for ID attributes: Exterior Wall and Interior Wall are (Wall.X), Window is (Window.X), Overhang, Fin, and AdjacentBuilding are (Shading.X), RoofSlab, FloorSlab, and BaseSlab are (Slab.X) in terms of ID attribute naming conventions. 

Using the existing schema, object type is “surfaceType”. Here’s the mapping from the schema to the above objects:

j.	Exterior Wall = ExteriorWall
k.	Interior Wall = InteriorWall
l.	Overhang = Shade
m.	Fin = Shade
n.	RoofSlab = Roof
o.	BaseSlab = SlabOnGrade
p.	FloorSlab = InteriorFloor
q.	AdjacentBuilding = Shade

Windows are openings. As above, opening ID attributes will be Window.1, Window.2, etc. openingType = FixedWindow

For geometry coordinate ordering, convention is all exterior walls start upper left-hand corner, counterclockwise when looking at the surface from the outside. Shading and Slab objects can have any starting vertice, but from there counterclockwise viewing it from above (or from the side for AdjacentBuildings and Fins). For interior walls, I believe you take upper left hand vertice, then counterclockwise from the perspective if either bounded space. Chien Si, is this correct?

Each space needs to have its bounding surface objects defined. One way to do this is look at the coordinates of the space floor geometry, then loop through surfaces that share 2 sequential vertices with the space, and you know it’s bounding. Floor object bounding has same coordinates as space floor coordinates, upper space boundary will have same coordinates as floor with the z-value increased by the FloorHeight value. Interior Walls will have common vertices with 2 spaces. If you have a slicker way of doing this using three.js, by all means use it.

Here are the required minimum objects for the gbXML file, see comment in red next to each:

•	(attribute) lengthUnit (1 max) BW-This will always be SI.
•	(attribute) temperatureUnit (1 max) BW-This will always be SI.
•	(attribute) volumeUnit (1 max) BW-This will always be SI.
•	(attribute) useSIUnitsForResults (1 max) BW-This will always be SI.
•	(attribute) version (1 max)
•	(child Element) Campus (1 max) BW-Just leave this as Campus.
o	(attribute) id (1 max) BW-Campus.1.
o	(child Element) Location (1 max) BW-Just put Minneapolis.
	child Element) ZipCode (1 max) BW-Just put 55305.
o	(child Element) Building (array with no maximum bound)
	(child Element) Area (1 max)
	(child Element) Space (array with no maximum bound
	(attribute) id (1 max) BW-Space.1, Space.2, etc.
	(child Element) SpaceBoundary (array with a minimum count of 4 with no maximum bound)
	(attribute) surfaceIdRef ** (1 max) BW-This will be Wall.X or Slab.X which is defined later.
	(child Element) Planar Geometry (1 max)
	(child Element) PolyLoop (1 max)
	(child Element) CartesianPoint (array with 3 min and no max bound)
	(child Element) Coordinate (array with min 2 and max 3)
	(Element) Area (1 max)
	(Element) Volume (1 max)
	(Element) PeopleNumber (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
	(Element) PeopleHeatGain (3 required) BW-Chien Si can we put null for this or are 3 integers required? If the latter, just put 1,1,1.
	(Element) LightPowerPerArea (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
	(Element) EquipPowerPerArea (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
	(Element) AirChangesPerHour (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
	(Element) InfiltrationFlow (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
o	(attribute) lightScheduleIdRef (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
o	(attribute) equipmentScheduleIdRef (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
o	(attribute) peopleScheduleIdRef (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
	(attribute) id (1 max, must be unique) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
	(attribute) buildingType (1 max) BW-Chien Si can we put null for this or is an integer required? If the latter, just put 1.
o	(child Element) Surface (an array with no maximum bound)
	(attribute) id (1 max) BW-Wall.1, Slab.1, Shading.1, etc.
	(attribute) surfaceType (1 max) BW-ExteriorWall, InteriorWall, Shade, Roof, SlabOnGrade, InteriorFloor.
	(attribute) constructionId (1 max) BW-Null or 1.
	(child Element) AdjacentSpaceId (an array that requires 1 minimum, 2 maximum elements) BW-Space.1, etc.  unless InteriorWall, then put 2 Space IDs.
	(child Element) RectangularGeometry (1 max)
	(child Element) Azimuth (1 max)
	(child Element) CartesianPoint (1 max) complete
	(child Element) Coordinate (array with min 2 and max 3)
	(child Element) Tilt (1 max)
	(child Element) Height (1 max)
	(child Element) Width (1 max)
	(element) PlanarGeometry) (1 max) complete
o	(child Element) Construction)(array with no maximum bound)
	(child Element) U-Value (1 max) BW-I believe the opening object specification will be within the bounding surface object (i.e. Wall or roof if skylight).

**Comment by bwelle** [comment]
Here I'm putting some content from space/zone integration issue and closing it. We can start making smaller issues off this main gbXML thread..
From @theo-armour 

A couple of days ago, I built a preliminary gbXML Schema Viewer:

[gbxml schema file viewer r1
](http://www.ladybug.tools/spider/read-gbxml/gbxml-schema-file-viewer/r1/gbxml-schema-file-viewer.html)
I hope that the Build Well project can link to the schema and use it a resource and for validation.

**Comment by bwelle** [comment]
From @theo-armour 

@chiensiTB

It would be great to come up with a features list and a roadmap.
Features list and roadmap [here](http://www.ladybug.tools/spider/#sandbox/build-well/README.md) but lagging bhind current reality:

Most of the information that you will need as per the request above is contained in the Surface object.
Thank you confirming my understanding.

If you go [here](http://www.ladybug.tools/spider/sandbox/build-well/):

In left menu: click: 'gbXML Export' then click 'getBuildingData'

A textarea should appear that contains the gbXML for the project. You may also export this data to a file, if you wish by clicking the button just below. You will note that 'Surface' is the only object exported (for the moment.) No openings yet. They - and other details - should start appearing soon.

The only other thing worth mentioning here is that all Surfaces have an Id attribute
What I need now are the rules for creating/naming the IDs and any the data required by Open Studio for all of the Shapes outlined by Ben elsewhere.

See you online later today

**Comment by chiensiTB** [comment]
@theo-armour , @bwelle  ~

You had asked a question about SpaceBoundary vs. Surface elements.  They are related to each other but a little different.  To be "official" gbXML and receive an official stamp of approval, you need Space Boundaries.  But for this exercise, just Surfaces will do fine.  In fact, an overwhelming number of gbxml producers never produce SpaceBoundary objects, and nearly all energy modeling converters consume only the Surface elements.

Everything else you show makes sense to me.  One note is floors and ceilings are interchangeable objects in gbxml.  there are some nuances to using one vs. the other, but this should not trip you up terribly at this point.

**Comment by bwelle** [comment]
Awesome. Let's jam.

**Comment by bwelle** [comment]
@theo-armour 
I am getting quite confused with the various versions and dev strategies for gbXML. Please help shed light on the following (anyone else feel free to jump in if you know)...

1. There's Mostapha's gbXML viewer, and the one you are doing for Mike, and the gbJSON viewer. Are the first 2 one in the same? 
2. All of them seem to use the same gbXML test files. However, when the test files are downloaded and imported into either OS desktop or FZK Viewer, none of them work and import fails. Why is this?
3. Mike, you said you like using FZK viewer to test gbXML files...which ones have you actually imported successfully?
4. When I take current Build Well gbXML exports, it doesn't open in FZK viewer like all the other "test cases", and some objects show up in OS desktop (error messages pertain to space assignments). Are you using Mostapha's/Mike's viewer for visual testing? Are you also using OS desktop for visual testing? It appears OS desktop does provide some useful error messages.
5. I have a gbXML file generated by Autodesk Vasari I used for my PhD [here](https://perkinswill.box.com/s/rwn6ewj5s79kvwvdb4h3aifpkdcu9qtv). It works perfectly in the OS desktop import, the FZK viewer, Mostaphas/Miki's viewer, and the gbJSON viewer (except in this one windows don't appear). What is it about this old gbXML file that makes it work everywhere where all the test files/Build Well gbXML files fail? Is there anything we can learn from this file?
6. In general, I just want to make sure all these gbXML related dev efforts are streamlined, and it would be great if somehow I could be assistance in debugging as both Chien Si and I have spent much time testing these files over various past projects (like when Chien Si built a "gbXML wrapper" for my PhD work) and I have also spent a large amount of time debugging IFC space boundary models (which should look exactly like the gbXML file geometry when using surfacic geometry. 
7. What is the incremental dev strategy you are following for Building Well? Is there someway I may be helpful in this process? Could we formalize dev/testing/debugging in some way so we can maximize each others knowledge?

Thanks!

**Comment by theo-armour** [comment]
@bwelle 

> I am getting quite confused with the various versions

Sorry for the confusion with the names. One evening after drinking too much I decided that numbers were boring and that revisions should have names instead of numbers. So far there's been releases titled 'harriman', 'Carmel' for Carmelsoft, 'Michal', and 'Mostapha'. I was going to name the next one 'Benjamin' but it's safer to go to R5. (you are already the third person who has raised an issue on this topic and it's not as if there are a lot of users out there. ;-)

> When I take current Build Well gbXML exports, it doesn't open in FZK viewer like all the other...

The Build Well R8 that's now on GitHub is from yesterday. That version created gbXML only partially being accepted for import by Open Studio

![image](https://user-images.githubusercontent.com/547626/31370509-4d753e50-ad40-11e7-9df0-a87631a8c808.png)

We are doing better today. As you can see above, R9 has moderately good 3D display of the L Shape in Open Studio. I still have not uploaded Build Well R9 and gbXML Viewer ~~Benjamin~~ R5 but should do so soon.

>  have a gbXML file generated by Autodesk Vasari I used for my PhD

Works just fine in gbXML Viewer R9. May I add it to the list of demo files?

> What is the incremental dev strategy you are following for Building Well gbXML export? Is there someway I may be helpful in this process?

I wish I knew. Until about four weeks ago, I had never looked at a gbXML file. I first heard about them in August. So I think what we has been accomplished in three or so weeks is quite amazing.

For the time being there's a lot of new code appearing - full of errors. I can see these as easily as anybody for the moment.

Things needed include:

* Spreadsheet with with area and dims for the various Shape calculations
* Deciding what Anton and ME do as opposed to what I do. For example, they should be preparing templates that follow your PDF specs and seeing if they can embed the relevant parts of the Build Well code into the Perkins Will dashboard thingy
* Updating the data for the scatter plot

**Comment by theo-armour** [comment]
@bwelle 

Here is a view of your PhD project in gbCML Viewer R5

![image](https://user-images.githubusercontent.com/547626/31371030-c402f97e-ad43-11e7-94e5-7a8df551918e.png)

**Comment by bwelle** [comment]
@theo-armour 
>We are doing better today. As you can see above, R9 has moderately good 3D display of the L Shape in >Open Studio. 

The image you posted doesn't the L coming in really at all. Random surfaces and little slivers that shouldn't be there. Are you uploading the correct image?

>Works just fine in gbXML Viewer R9. May I add it to the list of demo files?

I already posted it to demo file list, I think in the right place. If not in right, post where ever you want. I have many more test cases like this if you want them.

>I wish I knew. Until about four weeks ago, I had never looked at a gbXML file. I first heard about them >in August. So I think what we has been accomplished in three or so weeks is quite amazing.

It has been quite amazing! Didn't intend to suggest otherwise. But in recognition that you are new to gbXML, I just wanted to make sure you know that Chien Si and I are here to help any way, any time.:) For example, I can see an incorrect surface on the L-Shape on Build Well if I toggle a few things. Not sure if you've noticed it, but it's what is causing the sliver I believe.

>Spreadsheet with with area and dims for the various Shape calculations
>Deciding what Anton and ME do as opposed to what I do. For example, they should be preparing >templates that follow your PDF specs and seeing if they can embed the relevant parts of the Build Well >code into the Perkins Will dashboard thingy
>Updating the data for the scatter plot

Spreadsheet coming in 20 minutes. 

We have spent time for a while now on integration with ME. We have a meeting tomorrow morning to discuss how their first attempt to integrate your code with the Web GUI code on Pentaho (Angular) went for Site Context. Will update you once we learn how it went.

New simulations have been run. Anton is sending me new spreadsheet tomorrow to do a few checks on. then will update google docs sheet 1 with the new dataset, hopefully by early afternoon tomorrow.
